src/component.js

```diff
@@ -0,0 +1,146 @@
+/**
+ * @license
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Base class for all MDL components.
+ *
+ * @export
+ */
+class MaterialComponent {
+  /**
+   * Initialize component from a DOM node.
+   *
+   * @param {Element} root The element being upgraded.
+   */
+  constructor(root) {
+    this.root_ = root;
+
+    // Check if the root has the right class.
+    if (!root.classList.contains(this.constructor.classes_.ROOT)) {
+      throw new Error(`${this.constructor.classAsString} missing ` +
+          `${this.constructor.classes_.ROOT} class.`);
+    }
+  }
+
+  /**
+   * CSS classes used in this component.
+   *
+   * @protected
+   * @return {Object<string, string>}
+   */
+  static get classes_() {
+    // Empty in base class. Throw error if not correctly overriden.
+    throw new Error('Should have ROOT and JS keys with the style class names,' +
+      ' e.g. mdl-button and mdl-js-button.');
+  }
+
+  /**
+   * Number constants used in this component.
+   *
+   * @protected
+   * @return {Object<string, number>}
+   */
+  static get numbers_() {
+    // Empty in base class.
+    return {};
+  }
+
+  /**
+   * String constants used in this component.
+   *
+   * @protected
+   * @return {Object<string, string>}
+   */
+  static get strings_() {
+    // Empty in base class. Throw error if not correctly overriden.
+    throw new Error('Should have a CLASS_NAME key with the script class name,' +
+        ' e.g. MaterialButton.');
+  }
+
+  /**
+   * Return class name as a string. Useful for automation after obfuscation.
+   *
+   * @return {string}
+   * @suppress {missingProperties}
+   * @nocollapse
+   * @export
+   */
+  static get classAsString() {
+    return this.strings_.CLASS_NAME;
+  }
+
+  /**
+   * Initialize all self-managed instances under the given node.
+   * This will only initialize components that have specifically been marked
+   * for self-management.
+   *
+   * @param {Node} docRoot The node under which to look for components.
+   * @nocollapse
+   * @export
+   */
+  static initComponents(docRoot) {
+    var nodes = docRoot.querySelectorAll(
+        `.${this.classes_.ROOT}.${this.classes_.JS}`);
+    for (let i = 0; i < nodes.length; i++) {
```

Comment:
`NodeList`s are not arrays, so that's unsafe. From https://developer.mozilla.org/en/docs/Web/API/NodeList :

> Don't be tempted to use for...in or for each...in to enumerate the items in the list, since that will also enumerate the length and item properties of the NodeList and cause errors if your script assumes it only has to deal with element objects. Also, for..in is not guaranteed to visit the properties in any particular order.